File: Managed/UnrealSharp/UnrealSharp/Interop/FWorldDelegatesExporter.cs
using System.Runtime.InteropServices;
using UnrealSharp.Binds;
using UnrealSharp.Core;

namespace UnrealSharp.Interop;

public struct FWorldDelegates
{
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void FWorldCleanupEvent(IntPtr world, NativeBool sessionEnded, NativeBool cleanupResources);
}

[NativeCallbacks]
public static unsafe partial class FWorldDelegatesExporter
{
    public static delegate* unmanaged<IntPtr, out FDelegateHandle, void> BindOnWorldCleanup;
    public static delegate* unmanaged<FDelegateHandle, void> UnbindOnWorldCleanup;
}

File: Source/UnrealSharpCore/Private/Export/FWorldDelegatesExporter.cpp
#include "CSBindsManager.h"

DECLARE_UNREALSHARP_EXPORTER(FWorldDelegatesExporter)
{
	using FWorldCleanupEventDelegate = void(*)(UWorld*, bool, bool);
	
	void BindOnWorldCleanup(FWorldCleanupEventDelegate Delegate, FDelegateHandle* Handle)
	{
		*Handle = FWorldDelegates::OnWorldCleanup.AddLambda(Delegate);
	}

	void UnbindOnWorldCleanup(const FDelegateHandle Handle)
	{
		FWorldDelegates::OnWorldCleanup.Remove(Handle);
	}
	
	EXPORT_UNREALSHARP_FUNCTION(BindOnWorldCleanup)
	EXPORT_UNREALSHARP_FUNCTION(UnbindOnWorldCleanup)
}
